src/OptimalFilter.jl

```diff
@@ -0,0 +1,406 @@
+
+    for (num,(i,j)) in enumerate(zip(stations.ist, stations.jst))
+        values[num] = field[i,j]
+    end
+
+    return values
+
+end
+
+# Allocate and compute matrices that do not depend on time-dependent variables (height and observations).
+function init_offline_matrices(grid::Grid,
+                               grid_ext::Grid,
+                               stations::NamedTuple,
+                               noise_params::NamedTuple,
+                               obs_noise_std::T,
+                               fft_plan::FFTW.FFTWPlan,
+                               fft_plan!::FFTW.FFTWPlan,
+                               F::Type,
+                               ) where T
+
+    n1 = grid.nx * grid.ny # number of elements in original grid
+    n1_bar = grid_ext.nx * grid_ext.ny # number of elements in extended grid
+
+    C = complex(F)
+
+    matrices = OfflineMatrices(Vector{F}(undef, n1_bar),                   #rho_bar
+                               Matrix{F}(undef, n1, stations.nst),          #R12
+                               Matrix{F}(undef, stations.nst, n1_bar),      #R21_bar
+                               Matrix{F}(undef, stations.nst, stations.nst), #R22
+                               Matrix{F}(undef, stations.nst, stations.nst), #R22_inv
+                               Matrix{F}(undef, n1, stations.nst),          #R12_invR22
+                               Vector{F}(undef, n1_bar),                   #Lambda (diagonal elements)
+                               Matrix{C}(undef, stations.nst, n1_bar),      #K
+                               Matrix{F}(undef, stations.nst, stations.nst), #L
+                               Matrix{F}(undef, stations.nst, stations.nst), #mu20
+
+                               Matrix{F}(undef, grid.nx, grid.ny), #buf1
+                               Matrix{F}(undef, grid.nx, grid.ny)  #buf2
+                               )
+
+    first_column_covariance_grid!(matrices.rho_bar, grid, grid_ext, noise_params)
+    covariance_grid_stations!(matrices.R12, grid, stations, noise_params)
+    covariance_stations_extended_grid!(matrices.R21_bar, grid, grid_ext, stations, noise_params)
+    covariance_stations!(matrices.R22, grid, stations, noise_params, obs_noise_std)
+    matrices.R22_inv .= inv(matrices.R22)
+    matrices.R12_invR22 .= matrices.R12 * matrices.R22_inv
+
+    fourier_coeffs = Vector{C}(undef, n1_bar)
+    normalized_2d_fft!(fourier_coeffs, matrices.rho_bar, fft_plan, fft_plan!, grid_ext, inv)
+    matrices.Lambda .= sqrt(n1_bar) .* real.(fourier_coeffs)
+
+    WHbar_R12 = Matrix{C}(undef, n1_bar, stations.nst)
+    for i in 1:stations.nst
+        normalized_2d_fft!(@view(WHbar_R12[:,i]), @view(matrices.R21_bar[i,:]), fft_plan, fft_plan!, grid_ext)
+    end
+    KH = Diagonal(matrices.Lambda)^(-1/2)*WHbar_R12
+    matrices.K .= KH'
+
+    A = real.(matrices.R22 .- matrices.K*KH)
+
+    if ishermitian(A)
+        matrices.L .= cholesky(A).L
+    end
```

Comment:
Maybe we should do something like
```julia
    if A ≈ Hermitian(A)
        matrices.L .= cholesky(Hermitian(A)).L
    else
        error("R22 - K*KH is not hermitian! Increasing the grid size may help. See Dietrich and Newsam '96 for details.")
    end
```

Reply:
Tuomas mentioned in his email: "My suggestion is forcing the matrix to be hermitian when it is already approximately so, and reporting an error when it is not. Does this make sense to you, or do you have other suggestions?" 

I agree, give that the issue is indeed the one that you mention. But I have some comments about the "issue" itself. I do not think the issue is the particular matrix being Hermitian - I think it should be by definition (if I remember well from the Theory). If that recollection is correct, then if you get such a message "of not being Hermitian" as an error message then it *has* to be due to round-off errors. Continuing with my recollection, some issue that indeed arises with this Circulant Embedding method that the algorithm uses is that the matrix you are referring to can have negative eigenvalues - so the Cholesky decomposition method will give an error message. There is no easy way around this. What I mean, is that if one fixes the 2D domain, and the grid, then the only possible solution is some sort of "enlargement" of the Embedding, i.e. we are talking about adapting the core algorithm, so let's not go down this direction. From my understanding, the larger the 2D domain, the more likely that the matrix is positive definite, so in practice we just need to adjust our 2D domain - and grid - in a way that the matrix *is* positive definite, that is the practical solution.

Something I do not understand, is why should the above be related to getting different results at different executions of the algorithm under the same parameters. I mean, once the parameters are chosen, then either there is some issue with this matrix or there is no issue!!

Reply:
> Something I do not understand, is why should the above be related to getting different results at different executions of the algorithm under the same parameters.

Because if we don't explicitly set `matrices.L` here we access uninitialised memory.  In principle we could always set `matrices.L` to zero before here, but that's slightly slower than using an uninitialised array, which actually made debugging the issue (`matrices.L` not set correctly under certain conditions) much easier.

Reply:
Hmmm, I see - more or less. 
Is it something that is easy/difficult to fix?